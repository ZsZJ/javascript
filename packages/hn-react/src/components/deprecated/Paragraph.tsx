import PropTypes from 'prop-types';
import React from 'react';
import deprecate from 'util-deprecate';
import EntityMapper, { EntityMapperProps } from '../EntityMapper';

const Paragraph: React.StatelessComponent<ParagraphProps> = ({
  mapper,
  uuid,
  page,
  index,
  paragraphProps,
}) => (
  <EntityMapper
    mapper={mapper}
    uuid={uuid}
    entityProps={{ index, page, ...paragraphProps }}
  />
);

interface ParagraphProps {
  mapper: any;
  uuid: any;
  page?: object;
  index?: number;
  paragraphProps?: object;
}

Paragraph.propTypes = {
  index: PropTypes.number,
  mapper: PropTypes.oneOfType([PropTypes.shape({}), PropTypes.func]).isRequired,
  page: PropTypes.shape({}),
  paragraphProps: PropTypes.shape({}),
  uuid: PropTypes.string.isRequired,
};

Paragraph.defaultProps = {
  index: 0,
  page: undefined,
  paragraphProps: {},
};

export default deprecate(
  Paragraph,
  'The Paragraph component is deprecated, use the EntityMapper component instead',
);
