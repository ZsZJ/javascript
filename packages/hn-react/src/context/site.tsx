import { Site } from 'hn';
import React, { createContext, ReactNode } from 'react';
import globalSite from '../utils/site';

// Create a new context.
const { Provider, Consumer } = createContext(globalSite);

// The site provider is the same as the 'Provider', but changes 'site' to
// 'value'.
export const SiteProvider = ({
  site,
  children,
}: {
  site: Site;
  children: ReactNode;
}) => <Provider value={site} children={children} />;

// The consumer is exported as both Site and SiteConsumer. The SiteConsumer can
// be used if also the Site from 'hn' is imported, so the names don't overlap.
export { Consumer as Site, Consumer as SiteConsumer };

// There is also a HOC withSite that adds the site to the props of the
// component. Source of the HOC for Typescript:
// https://stackoverflow.com/a/50613946/1907875
export interface InjectedSiteProps {
  site: Site;
}

export function withSite<P extends InjectedSiteProps>(
  Component: React.ComponentType<P>,
) {
  return function ComponentWithSite(
    props: Pick<P, Exclude<keyof P, keyof InjectedSiteProps>>,
  ) {
    return (
      <Consumer>{theme => <Component {...props} site={theme} />}</Consumer>
    );
  };
}
