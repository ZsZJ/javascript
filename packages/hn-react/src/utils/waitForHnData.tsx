import * as PropTypes from 'prop-types';
import * as React from 'react';

const asyncBootstrapper = require('react-async-bootstrapper'); // tslint:disable-line:no-var-requires

export default async function(children) {
  const context = {
    state: {
      drupalPage: null,
      entities: [],
    },
  };

  class DrupalPageContextProvider extends React.Component {
    static childContextTypes = {
      hnContext: PropTypes.object,
    };

    getChildContext() {
      return { hnContext: context };
    }

    render() {
      return React.Children.only(this.props.children);
    }
  }

  const drupalContextProvider = (
    <DrupalPageContextProvider>{children}</DrupalPageContextProvider>
  );

  await asyncBootstrapper(drupalContextProvider, {
    componentWillUnmount: true,
  });

  return drupalContextProvider;
}
