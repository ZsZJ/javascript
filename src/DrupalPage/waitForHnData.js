import React, { Component } from 'react';
import asyncBootstrapper from 'react-async-bootstrapper'
import PropTypes from 'prop-types';

export default async function (children) {

  const context = {};

  class DrupalPageContextProvider extends Component {
    getChildContext() {
      return {hnContext: context};
    }

    render() {
      return React.Children.only(this.props.children);
    }
  }

  DrupalPageContextProvider.childContextTypes = {
    hnContext: PropTypes.object,
  };

  const drupalContextProvider = (
    <DrupalPageContextProvider>
      {children}
    </DrupalPageContextProvider>
  );

  await asyncBootstrapper(drupalContextProvider);

  return drupalContextProvider;

};
