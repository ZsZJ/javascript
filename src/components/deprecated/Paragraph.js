import React from 'react';
import EntityMapper from '../EntityMapper';
import PropTypes from 'prop-types';
import deprecate from 'util-deprecate';

function Paragraph ({ mapper, uuid, page, index, paragraphProps }) {
  return <EntityMapper
    mapper={mapper}
    uuid={uuid}
    entityProps={{index, page, ...paragraphProps}}
  />
}

Paragraph.propTypes = {
  mapper: PropTypes.oneOfType([
    PropTypes.shape(),
    PropTypes.func,
  ]).isRequired,
  uuid: PropTypes.string.isRequired,
  page: PropTypes.shape({}),
  index: PropTypes.number,
  paragraphProps: PropTypes.shape(),
};

Paragraph.defaultProps = {
  paragraphProps: {},
  index: 0,
  page: undefined,
};

export default deprecate(Paragraph, 'The Paragraph component is deprecated, use the EntityMapper component instead');
