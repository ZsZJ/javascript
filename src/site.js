import { Site } from 'hn';
import getNested from 'get-nested';

const site = new Site({ url: '' });

if(getNested(() => window.__NEXT_DATA__.HN_DATA)) {
  site.data = window.__NEXT_DATA__.HN_DATA;
  site.url = window.__NEXT_DATA__.HN_URL || site.url;
  site.pagesLoading = {
    [window.location.pathname]: Promise.resolve(),
  };
}

export default site;
