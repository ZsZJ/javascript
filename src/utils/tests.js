import React from 'react';

const mapperComponent = jest.fn((props) => {
  return <div className="MyCustomMapperComponent" {...props} />
});

export const mapper = {
  unique_type__unique_bundle: mapperComponent
};

export const uuid = 'unique-uuid';

export const entity = {
  __hn: {
    entity: {
      type: 'unique_type',
      bundle: 'unique_bundle'
    }
  }
};

export function mockSite() {
  class SiteMock {
    getData = jest.fn(() => {
      return entity;
    });
    getPage = jest.fn(async () => {
      return uuid;
    })
  }
  return new SiteMock();
}
