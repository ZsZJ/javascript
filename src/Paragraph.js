import React from 'react';
import PropTypes from 'prop-types';
import site from './site';

const Paragraph = ({ mapper, uuid, page, index, paragraphProps }) => {
  const paragraph = site.getData(uuid);

  if(!paragraph) {
    return null;
  }

  const bundle = `${paragraph.__hn.entity.type}__${paragraph.__hn.entity.bundle}`;

  const Component = (typeof mapper === 'object')
    ? mapper[bundle]
    : typeof mapper === 'function'
      ? mapper(page, bundle)
      : null;

  if(!Component) {
    return null;
  }

  return (
    <Component
      bundle={bundle}
      page={page}
      paragraph={paragraph}
      index={index}
      {...paragraphProps}
    />
  );
};

Paragraph.propTypes = {
  mapper: PropTypes.oneOfType([
    PropTypes.shape(),
    PropTypes.func,
  ]).isRequired,
  uuid: PropTypes.string.isRequired,
  page: PropTypes.shape({}),
  index: PropTypes.number.isRequired,
  paragraphProps: PropTypes.shape(),
};

Paragraph.defaultProps = {
  paragraphProps: {}
};

export default Paragraph;
