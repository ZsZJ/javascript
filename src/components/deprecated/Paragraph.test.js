import Paragraph from './Paragraph';
import React from 'react';
import renderer from 'react-test-renderer';
import site from '../../utils/site';
import waitForHnData from '../../utils/waitForHnData';
import { mapper, uuid } from '../../utils/tests';

jest.mock('../../utils/site', () => {
  return require('../../utils/tests').mockSite();
});

jest.mock('util-deprecate', () => jest.fn((func) => func));
console.log = console.warn = console.error = jest.fn((message) => { throw new Error(message); });

beforeEach(() => {
  site.getData.mockRestore();
});

describe('Paragraph', async () => {

  test('with required props', async () => {
    const component = await waitForHnData(
      <Paragraph
        mapper={mapper}
        uuid={uuid}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();
  });

  test('with all props', async () => {
    const component = await waitForHnData(
      <Paragraph
        mapper={mapper}
        paragraphProps={{
          testProp: 'testPropValue'
        }}
        uuid={uuid}
        page={{'pageTest': true}}
        index={15}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();
  });
});
