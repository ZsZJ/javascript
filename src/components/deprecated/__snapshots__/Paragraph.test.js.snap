// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph with all props 1`] = `
<div
  bundle="unique_type__unique_bundle"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  index={15}
  page={
    Object {
      "pageTest": true,
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  testProp="testPropValue"
/>
`;

exports[`Paragraph with required props 1`] = `
<div
  bundle="unique_type__unique_bundle"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  index={0}
  page={undefined}
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
/>
`;
